pallets/nfts/src/features/approvals.rs

```diff
@@ -165,10 +179,197 @@ impl<T: Config<I>, I: 'static> Pallet<T, I> {
+				let deposit_required = T::CollectionApprovalDeposit::get();
+				let mut current_deposit = match maybe_approval.take() {
+					Some((_, deposit)) => deposit,
+					None => Zero::zero(),
+				};
+
+				if current_deposit < deposit_required {
+					T::Currency::reserve(&origin, deposit_required - current_deposit)?;
+					current_deposit = deposit_required;
+				}
+				*maybe_approval = Some((deadline, current_deposit));
+				Ok(())
+			},
+		)?;
+
+		Self::deposit_event(Event::TransferApproved {
+			collection,
+			item: None,
+			owner: origin,
+			delegate,
+			deadline,
+		});
+
+		Ok(())
+	}
+
+	/// Cancels the transfer of items in the collection that owned by the origin to a delegate.
+	///
+	/// This function is used to cancel the approval for the transfer of items in the `collection`
+	/// that owned by the `origin` to a `delegate`. After canceling the approval, the function
+	/// returns the `origin` back the deposited fund and emits the `ApprovalCancelled` event.
+	///
+	/// - `origin`: The account grants permission to cancel the transfer.
+	/// - `collection`: The identifier of the collection.
+	/// - `delegate`: The account that was previously allowed to take control of items in the
+	///   collection that owned by the origin.
+	pub(crate) fn do_cancel_collection_approval(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		delegate: T::AccountId,
+	) -> DispatchResult {
+		let (_, deposit) = CollectionApprovals::<T, I>::take((&collection, &origin, &delegate))
+			.ok_or(Error::<T, I>::UnknownCollection)?;
+
+		T::Currency::unreserve(&origin, deposit);
+
+		Self::deposit_event(Event::ApprovalCancelled {
+			collection,
+			owner: origin,
+			item: None,
+			delegate,
+		});
+
+		Ok(())
+	}
+
+	/// Clears all collection approvals.
+	///
+	/// This function is used to clear all approvals to transfer items in the `collection` that
+	/// owned by the `origin` to a `delegate`. After clearing all approvals, the function returns
+	/// the `origin` back the deposited fund of each collection approval and emits the
+	/// `AllApprovalsCancelled` event.
+	///
+	/// - `origin`: The account grants permission to clear the transfer.
+	/// - `collection`: The collection ID containing the item.
+	/// - `witness_approvals`: Information on the collection approvals cleared. This must be
+	///   correct.
+	pub(crate) fn do_clear_all_collection_approvals(
+		origin: T::AccountId,
+		collection: T::CollectionId,
+		witness_approvals: u32,
+	) -> DispatchResult {
+		let mut removed_approvals: u32 = 0;
+		// Iterate and remove each collection approval, return the deposited fund back to the
+		// `origin`.
+		for (_, (_, deposit)) in CollectionApprovals::<T, I>::drain_prefix((collection, &origin)) {
+			ensure!(removed_approvals < witness_approvals, Error::<T, I>::BadWitness);
+			T::Currency::unreserve(&origin, deposit);
+			removed_approvals.saturating_inc();
+		}
+		ensure!(removed_approvals == witness_approvals, Error::<T, I>::BadWitness);
```

Comment:
Should be checked how this is done in other places in the pallet but if `removed_approvals` is larger, we basically already did the actions and the pre-dispatched amount is invalid. This is an attack vector

Reply:
We have a code at line 305
```rs
ensure!(removed_approvals < witness_approvals, Error::<T, I>::BadWitness);
```
To make sure that this never go over the `witness_approvals`

Reply:
Missed that, thank you

Reply:
Curious to hear other opinions about this comparision. Do we still need it? Instead of calling it `witness_approvals`, we can call it `max_removed_approvals` or something similar and allow the caller to remove the approvals up to the limit, return the `removed_approvals` for post_dispatch` weight. This approach will simplify the UX and still calculate the weight correctly.

Reply:
Lets call it limit